### ISSUE TITLE
[Request] Stretch to full screen

### ISSUE THREAD

**Issue by raven02** [yuzu-emu/yuzu#2427]
Override aspect ratio and stretches the image to the full display area

**Comment by DarkLordZach** [comment]
Could you elaborate?

**Comment by raven02** [comment]
override aspect ration and stretch the image to the full display area

**Comment by DarkLordZach** [comment]
The default aspect ratio is the switch's (which is 16:9). Stretching the image would deteriorate the image quality substantially (depending on how far from 16:9 your monitor aspect ratio is). Most of our userbase uses yuzu on a 1080p 1920x1080 monitor (or a down scaling of that) -- which is 16:9.

**Comment by raven02** [comment]
Though not much users are using 21:9 monitor here and would be nice to have it stretch to full screen to get rid of the two sides black bar .

**Comment by niko1point0** [comment]
I’m not sure how this would work. Aspect ratio is part of perspective projection, which is inside one matrix (out of many) that are passed to shaders. How can the emulator programmers know exactly which matrix in the shader contains the aspect ratio that needs changing?

One possibility is to hook a function like glm::perspective, and override it, but then any game that doesn’t use that exact function for projection matrix generation will still be locked to 16:9

Please correct me if I’m wrong

**Comment by Relsig** [comment]
TL;DR : Correct me if I'm wrong but...
It seems to me @niko1point0 is talking about changing aspect ratios internally whereas @raven02 seems talking about stretching the final product to fit the screen size.

The long version :

I don't follow yuzu or citra well enough to know how or where you've set up the display area code but the latter is "probably" not a substantial change. The former would be hackish no matter how it was implemented.

I used quotes because I follow yuzu to appease my interest in emulation and how technology works. I don't actually use yuzu myself. However, If you can use the emulator windowed and resize that window arbitrarily you've dictated how that happens somewhere in your code, otherwise it would fill the whole window and there would be no black bars. 

If you refer to the pull request raven mentioned you'll see it added a config option that simply disables some code when true, reverting the renderer to a more natural state.

As a preferrer of stretched graphics instead of black bars I wish you the best of luck should you choose to look for a solution.

**Comment by niko1point0** [comment]
If the image is simply stretched, the image quality might not be what you’re expecting. As far as I know, if you stretch a 16:9 image to fit a 21:9, without internally changing aspect ratio, then all you’re doing is making every object in the scene look shorter and wider

For example, look at PS1 games that were released on PSP. There were display modes to have big black bars, and there was a mode to stretch, which did not internally change aspect ratio, and it doesn’t look that good to me

**Comment by EliEron** [comment]
While I can't speak for Raven and Relsig, I'm fairly certain they do understand the effect stretching will have on the image quality. There is a somewhat significant group of people that don't mind the distortion that stretching causes, and in fact prefer it over black bars. That's why the option to ignore aspect ratios is present in most Media Players, Emulators and Televisions/Monitors in the first place.

I'm not personally a fan of stretching either, but I don't see much of an issue with adding it as an option. Also enabling screen stretching is a prerequisites step when using widescreen cheats on pretty much all emulators. So while it wouldn't be automatic, once stretching is supported somebody could come along and create hacks or cheats that corrects the aspect ratio for specific games.

**Comment by Relsig** [comment]
@EliEron it seems like you do understand raven's intent. Using cheat engine you can modify the aspect ratio with a high probability of success, without bulking up the emulator's code. Unfortunately being able to fill the display area is a requirement to make that modification work. This is the route Raven will most likely take.

You are also correct in assuming a substantial minority of people do prefer warped objects over black bars, myself included. The option shouldn't be terribly hard to implement for someone who understands the internal workings of the display area code, it's simply a matter of finding the right dev and waiting for them to have some spare time.

**Comment by ReinUsesLisp** [comment]
Modern GPUs do not manage projection or modelview matrices anymore. In particular, Switch's Nvidia Maxwell GPU most common float-point instructions don't even handle vectors (I don't consider half-float SIMD instructions common). We can't know what is a color, what is a normal, what's a projection matrix or what's a skeleton weight: in the shader it's all memory reads, stores and operations. Emulators for older consoles approach to change aspect ratios and resolution changes are not applyable to Switch emulation.

What other modern console emulators have done (e.g. Cemu) is to have a relatively stable interface for modders to use that manually change shaders and texture layouts on a per-game basis. This way game's internal resolution can be changed with some work. Once the internal resolution problem is solved, presenting to a different screen size as a final step is an easy change to do.

One little detail I'd like to clarify: `glm::` are very likely going to be inlined by whatever compiler game developers on deployment (probably clang), so those functions are lost.

In the meantime, I'm against stretching the final image. If black bars really bother, remember the monitor showing yuzu very likely has borders too.

**Comment by Relsig** [comment]
I think raven understands that changing aspect ratios internally is a big ask. His request seems to simply be, as you said, the option (in yuzu's settings) to present to a different screen size. I believe this is why he chose the title he chose.

Given the amazing developers working on Yuzu I'm sure in the distant future an interface to change textures and shaders will happen but I don't believe that is what this request was made for.

**Comment by Margen67** [comment]
This is possible in Citra so I'm not sure why an option for this isn't present in yuzu.

Ultrawide hacks are a thing (pixelated/blurry since resolution scaling was removed):
![](https://cdn.discordapp.com/attachments/495758692495523854/645468396598853642/uw.png)
https://ptb.discordapp.com/channels/398318088170242053/402001453381713920/590289681170563082
![](https://cdn.discordapp.com/attachments/402001453381713920/590289663949013013/unknown.png)
![smo](https://user-images.githubusercontent.com/3462541/69003396-3edd9300-08b6-11ea-8f11-b9c716f52673.png)

**Comment by Margen67** [comment]
Update: For anyone that can build yuzu and wants to have the screen stretched, here's how:
Change <https://github.com/yuzu-emu/yuzu/blob/master/src/core/frontend/framebuffer_layout.cpp#L30-#L31>
```
const float emulation_aspect_ratio{static_cast<float>(ScreenUndocked::Height) /
                                       ScreenUndocked::Width};
```
to
```
const float emulation_aspect_ratio = static_cast<float>(height) / width;
```
![](https://cdn.discordapp.com/attachments/495758692495523854/671326804165918750/unknown.png)

**Comment by MiShi-Official** [comment]
> This is possible in Citra so I'm not sure why an option for this isn't present in yuzu.
> 
> Ultrawide hacks are a thing (pixelated/blurry since resolution scaling was removed): ![](https://camo.githubusercontent.com/0965696dd803c4b26c12b208b8985acf471febfaf3cad6cd8610d4b45cc6df9e/68747470733a2f2f63646e2e646973636f72646170702e636f6d2f6174746163686d656e74732f3439353735383639323439353532333835342f3634353436383339363539383835333634322f75772e706e67) https://ptb.discordapp.com/channels/398318088170242053/402001453381713920/590289681170563082 ![](https://camo.githubusercontent.com/5d3e163dcfed0c09cfaa32e4fa84faf4a64e614ade50e68aff3e2848b65b187b/68747470733a2f2f63646e2e646973636f72646170702e636f6d2f6174746163686d656e74732f3430323030313435333338313731333932302f3539303238393636333934393031333031332f756e6b6e6f776e2e706e67) ![smo](https://user-images.githubusercontent.com/3462541/69003396-3edd9300-08b6-11ea-8f11-b9c716f52673.png)

Hey saw this and it looks amazing, what program did you use to achieve 21:9, if that is lengthy would you be able to upload the cheat for it? appreciate the help!
